### PR TITLE
Backport script loader: enqueue stored block supports styles (refresh)

### PR DESCRIPTION
**Pull request by ockham**
🎼 _So Fresh, So Clean_ 🎶 

This is a fork of @ramonjd's https://github.com/WordPress/wordpress-develop/pull/3218, to rebase it (now that https://github.com/WordPress/wordpress-develop/pull/3199 has been merged), and to address [feedback by](https://github.com/WordPress/wordpress-develop/pull/3218#discussion_r971632145) @gziolo.

Original PR desc:

> Backporting changes to script-loader that enqueue block support styles stored by the style engine/
> 
> See tracking issue: [WordPress/gutenberg#43440](https://github.com/WordPress/gutenberg/issues/43440)
> 
> ### Backporting the following changes
> * [Style engine: enqueue block supports styles in Gutenberg gutenberg#42880](https://github.com/WordPress/gutenberg/pull/42880)

(https://github.com/WordPress/gutenberg/pull/4288 has testing instructions.)

Trac ticket: https://core.trac.wordpress.org/ticket/56467

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/theme/wpGetGlobalStylesheet.php [hunk 1]
```diff
@@ -54,6 +54,18 @@ public function tear_down() {
 		parent::tear_down();
 	}
 
+	/**
+	 * Cleans up global scope.
+	 *
+	 * @global WP_Scripts $wp_scripts
```
Comment (gziolo) [inline tests/phpunit/tests/theme/wpGetGlobalStylesheet.php]: Is it used `$wp_scripts` here?

Comment (ockham) [reply tests/phpunit/tests/theme/wpGetGlobalStylesheet.php]: Ah, copy-pasta! Good catch 😅  Only `$wp_styles` it seems. I'll remove `$wp_scripts`.

**Comment by ockham** [comment]
Unit tests are failing 😕 

```
1) Tests_Theme_wpGetGlobalStylesheet::test_should_enqueue_stored_styles
Registered styles with handle of "core-block-supports" do not match expected value from Style Engine store.
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => '.saruman{color:white;height:100px;border-style:solid;}'
+    0 => '/**\n
+     * Core styles: block-supports\n
+     */\n
+    .saruman {\n
+    	color: white;\n
+    	height: 100px;\n
+    	border-style: solid;\n
+    }\n
+    '
```

Edit: Looks like comments and whitespace aren't removed as expected 🤔

**Comment by ockham** [comment]
A similar issue seems to be present over at [#3237 (comment)](https://github.com/WordPress/wordpress-develop/pull/3237#issuecomment-1247405947) and https://github.com/WordPress/wordpress-develop/pull/3204#issuecomment-1248152841.

**Comment by andrewserong** [comment]
> A similar issue seems to be present over at https://github.com/WordPress/wordpress-develop/pull/3237#issuecomment-1247405947 and https://github.com/WordPress/wordpress-develop/pull/3204#issuecomment-1248152841.

I believe the test failure with this one is slightly different to those other two PRs. For the other two PRs, the inline styles tests likely need to be updated to either expect no space between the colon and the value (since the style engine intentionally switched inline style output to remove the space), or for the tests to be more permissive of spaces (i.e. remove spaces between property + colon and value before doing the comparison).

For this one, I think the issue is that (inferring from the test failures) the tests in core appear to be run with `SCRIPT_DEBUG` set, which switches on prettify for the stored styles output on the followings lines: https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/style-engine/class-wp-style-engine-processor.php#L104, and https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/script-loader.php#L2990

As far as I can tell, for these tests, the `SCRIPT_DEBUG` value is determined here: https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/default-constants.php#L108 (if I switch that line to `$develop_src = false;` then the tests pass for me locally).

So, what's the best fix for these particular tests? Given that `SCRIPT_DEBUG` appears to be set in `default-constants.php` based on whether or not the current WP version has `-src`, I was wondering if we should make the tests factor in whether `SCRIPT_DEBUG` is set to determine which value to expect.

I've opened up (yet another 😅) fork of this changeset in #3262 to try that out. But feel free to close that / copy + paste anything from there if it's helpful.

**Comment by ramonjd** [comment]
> So, what's the best fix for these particular tests? Given that SCRIPT_DEBUG appears to be set in default-constants.php based on whether or not the current WP version has -src, I was wondering if we should make the tests factor in whether SCRIPT_DEBUG is set to determine which value to expect.

Thanks for looking into this folks. Sorry I was out of action last week.

Yeah, while I was running these tests locally `SCRIPT_DEBUG` was activated and I didn't get time to go deeper.

There was a recent change to `wp_style_engine_get_stylesheet_from_context` however that adds an options arg, through which we can set `prettify => true/false`. Maybe we could update the function signature of `wp_enqueue_stored_styles()` to be able to pass that down?

Would that be kosher?


> I've opened up (yet another 😅) fork of this changeset in https://github.com/WordPress/wordpress-develop/pull/3262 to try that out. But feel free to close that / copy + paste anything from there if it's helpful.

Thank you @andrewserong !!!

**Comment by andrewserong** [comment]
> Maybe we could update the function signature of wp_enqueue_stored_styles() to be able to pass that down?
> Would that be kosher?

Passing `options` to that function sounds like a good idea to me, given that we're supporting `options` args for most of the public style engine methods, and it'd make the a bit more explicit (e.g. we could test both prettified and non-prettified if we wanted to) 👍

**Comment by ramonjd** [comment]
> Passing options to that function sounds like a good idea to me, given that we're supporting options args for most of the public style engine methods, and it'd make the a bit more explicit (e.g. we could test both prettified and non-prettified if we wanted to)

Okey dokey! I'll update this PR, but let's keep https://github.com/WordPress/wordpress-develop/pull/3262 just in case there are doubts. 🙇

File: tests/phpunit/tests/theme/wpGetGlobalStylesheet.php [hunk 2]
```diff
@@ -199,4 +210,64 @@ public function test_variables_in_classic_theme_with_presets_using_defaults() {
...
+		wp_style_engine_get_stylesheet_from_css_rules(
+			$my_styles_to_enqueue,
+			array(
+				'context' => 'my-styles',
+			)
+		);
+
+		wp_enqueue_stored_styles();
+
+		$this->assertEquals(
```
Comment (ramonjd) [inline tests/phpunit/tests/theme/wpGetGlobalStylesheet.php:261]: This is a hangover from my [original PR](https://github.com/WordPress/wordpress-develop/pull/3218). 

I think it's better to use `assertSame` here for string comparisons and `assertEquals` for objects etc

```suggestion
		$this->assertSame(
```

File: tests/phpunit/tests/theme/wpGetGlobalStylesheet.php [hunk 3]
```diff
@@ -199,4 +210,64 @@ public function test_variables_in_classic_theme_with_presets_using_defaults() {
...
+			array( '.saruman{color:white;height:100px;border-style:solid;}' ),
+			wp_styles()->registered['core-block-supports']->extra['after'],
+			'Registered styles with handle of "core-block-supports" do not match expected value from Style Engine store.'
+		);
+
+		$this->assertEquals(
```
Comment (ramonjd) [inline tests/phpunit/tests/theme/wpGetGlobalStylesheet.php:267]: As above.

```suggestion
		$this->assertSame(
```

File: tests/phpunit/tests/theme/wpGetGlobalStylesheet.php [hunk 4]
```diff
@@ -199,4 +210,64 @@ public function test_variables_in_classic_theme_with_presets_using_defaults() {
...
+				'selector'     => '.gandalf',
+				'declarations' => array(
+					'color'        => 'grey',
+					'height'       => '90px',
+					'border-style' => 'dotted',
+				),
+			),
+		);
+
+		// Enqueue some other styles.
```
Comment (ramonjd) [inline tests/phpunit/tests/theme/wpGetGlobalStylesheet.php:251]: I forgot to change this verb tense from imperative to third person present

```suggestion
		// Enqueues some other styles.
```

**Comment by ramonjd** [comment]
> There was a recent change to wp_style_engine_get_stylesheet_from_context however that adds an options arg, through which we can set prettify => true/false. Maybe we could update the function signature of wp_enqueue_stored_styles() to be able to pass that down?

@ockham I've added a Gutenberg PR with the proposed changes mentioned in the [comment above](https://github.com/WordPress/wordpress-develop/pull/3259#issuecomment-1250403735).

https://github.com/WordPress/gutenberg/pull/44248

Happy to refork this PR with the changes if you don't have the bandwidth to port them across to this PR.

**Comment by ramonjd** [comment]
> Happy to refork this PR with the changes if you don't have the bandwidth to port them across to this PR.

Update PR here: https://github.com/WordPress/wordpress-develop/pull/3273 

If you're happy with that, we can close this PR

**Comment by ockham** [comment]
Thank you @ramonjd! Closing in favor of #3273.

**Pull request closed without merging**